File: arena-rpg.java/test/PersonagemTest.java
import static org.junit.Assert.*;

import org.junit.Before;
import org.junit.Test;
import org.mockito.Mock;

import static org.mockito.Mockito.*;


public class PersonagemTest {

    private Personagem leonidas, hulk;
    @Mock private Arma espada;
    
    @Before
    public void setUp() throws Exception {
        espada = mock(Arma.class);
        leonidas = new Personagem("Leonidas", espada, 200);
        hulk = new Personagem("Hulk", null, 800);
    }

    @Test
    public void deveriaEstarVivo() {
        assertTrue(leonidas.estaVivo());
    }
    
    @Test 
    public void deveriaSaberQualADistanciaAteOutroPersonagem() {
    	leonidas.movePara(1, 2);
    	hulk.movePara(4, 2);
    	assertEquals("quando no ponto 1,2 e outro personagem no ponto 4,2, deveria saber que a distancia até ele é de 3", 
    			leonidas.distanciaAte(hulk), 3);
    	
    	leonidas.movePara(4, 2);
    	hulk.movePara(1, 2);
    	assertEquals("quando no ponto 4,2 e outro personagem no ponto 1,2, deveria saber que a distancia até ele é de 3", 
    			leonidas.distanciaAte(hulk), 3);

    	leonidas.movePara(1, 1);
    	hulk.movePara(7, 10);
    	assertEquals("quando no ponto 1,1 e outro personagem no ponto 7,10, deveria saber que a distancia até ele é de 10", 
    			leonidas.distanciaAte(hulk), 10);
    }
    
    @Test
    public void deveriaAtacarOutroPersonagem() {
    	when(espada.getDano()).thenReturn(20);
        leonidas.ataca(hulk);
        
        assertEquals(hulk.getHp(), 780);
    }
    
    @Test
    public void deveriaNaoEstarArmado() {
        assertFalse(hulk.estaArmado());
    }
    
    @Test
    public void deveriaEstarAtingivel() {
        leonidas.movePara(1, 2);

        Personagem chuck = new Personagem("Chuck Noris", null, 1);
        chuck.movePara(1, 1);
        
        when(espada.getDistancia()).thenReturn(1);
        
        assertTrue(leonidas.estaAtingivel(chuck));
    }
    
    @Test
	public void quandoDesarmadoDeveriaSaberQueOponenteNaoEstaAtingivel() {
		hulk.movePara(1, 2);

		Personagem chuck = new Personagem("Chuck Noris", null, 1);
		chuck.movePara(1, 1);

		assertFalse(
				"quando desarmado, deveria saber que oponente não está atingível",
				hulk.estaAtingivel(chuck));
	}
    
    @Test
    public void deveriaNaoEstaAtingivel() {
        leonidas.movePara(4, 4);

        Personagem chuck = new Personagem("Chuck Noris", null, 1);
        chuck.movePara(1, 1);
        
        assertFalse(leonidas.estaAtingivel(chuck));
    }
    
    @Test
    public void deveriaMoverPontoAoOutro() {
        leonidas.movePara(1, 1);
        leonidas.movePara(4, 5);

        assertEquals(4, leonidas.getX());
        assertEquals(5, leonidas.getY());
    }
    

}

File: arena-rpg.elixir/test/personagem_test.exs
defmodule PersonagemTest do
  use ExUnit.Case
  import Personagem

  setup do
    { :ok, arma: %{nome: "espada", distancia: 1, dano: 20} }
  end

  setup context do
    { :ok,
      [ leonidas: %Personagem{nome: "Leonidas", arma: context[:arma], hp: 200},
        hulk: %Personagem{nome: "Hulk", arma: nil, hp: 800}
      ]
    }
  end

  test "deveria criar um personagem", context do
    leonidas = context[:leonidas]
    assert leonidas.nome == "Leonidas"
    assert leonidas.hp == 200 
    assert leonidas.arma == context[:arma]
  end

  test "deveria estar vivo", context do
    assert esta_vivo(context[:leonidas])
  end

  test "deveria atacar outro personagem quando armado", context do
    vitima = ataca context[:leonidas], context[:hulk]
    assert vitima.hp == 780
  end

  test "deveria tentar atacar outro personagem quando desarmado", context  do
    atacante = %{ context[:leonidas] | arma: nil }
    assert_raise RuntimeError, "atacante necessita de uma arma", fn ->
      ataca atacante, context[:hulk]
    end
  end

  test "deveria estar armado", context do
    assert esta_armado?(context[:leonidas]) 
  end

  test "deveria não estar armado", context do 
    assert esta_armado?(context[:hulk]) == false
  end
  
  test "deveria estar atingível quando armado", context do
    atacante = context[:leonidas]
    atacante = move_para atacante, 1, 2
    vitima = %{ nome: "Chuck Noris", x: 1, y: 1 }

    assert esta_atingivel?(atacante, vitima)
  end

  test "deveria não estar atingível quando desarmado", context do
    atacante = %{ context[:leonidas] | arma: nil } 
    atacante = move_para atacante, 1, 2
    vitima = %{ nome: "Chuck Noris", x: 1, y: 1 }

    assert_raise RuntimeError, "atacante necessita de uma arma", fn -> 
      esta_atingivel?(atacante, vitima)
    end
  end

  test "deveria se mover de um ponto a outro", context do
    leonidas = context[:leonidas]
    leonidas = move_para leonidas, 1, 1
    leonidas = move_para leonidas, 4, 5

    assert leonidas.x == 4
    assert leonidas.y == 5
  end
end
